From 51f580c279440e02d561a7a0e2483c98421c2163 Mon Sep 17 00:00:00 2001
From: Andrea Arcangeli <aarcange@redhat.com>
Date: Tue, 9 Sep 2014 01:16:28 -0400
Subject: [mm] hugetlbfs: Add some VM_BUG_ON()s to catch non-hugetlbfs pages

Message-id: <1410225399-8094-12-git-send-email-aarcange@redhat.com>
Patchwork-id: 93358
O-Subject: [RHEL7.1 PATCH 11/22] mm: hugetlbfs: Add some VM_BUG_ON()s to catch non-hugetlbfs pages
Bugzilla: 1135506
RH-Acked-by: Rik van Riel <riel@redhat.com>
RH-Acked-by: Larry Woodman <lwoodman@redhat.com>
RH-Acked-by: Rafael Aquini <aquini@redhat.com>

From: Dave Hansen <dave.hansen@linux.intel.com>

Dave Jiang reported that he was seeing oopses when running NUMA systems
and default_hugepagesz=1G.  I traced the issue down to
migrate_page_copy() trying to use the same code for hugetlb pages and
transparent hugepages.  It should not have been trying to pass thp pages
in there.

So, add some VM_BUG_ON()s for the next hapless VM developer that tries
the same thing.

Signed-off-by: Dave Hansen <dave.hansen@linux.intel.com>
Reviewed-by: Naoya Horiguchi <n-horiguchi@ah.jp.nec.com>
Tested-by: Dave Jiang <dave.jiang@intel.com>
Acked-by: Mel Gorman <mgorman@suse.de>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>
Signed-off-by: Andrea Arcangeli <aarcange@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/include/linux/hugetlb.h b/include/linux/hugetlb.h
index 5fb0244..0ed99a9 100644
--- a/include/linux/hugetlb.h
+++ b/include/linux/hugetlb.h
@@ -346,6 +346,7 @@ static inline pte_t arch_make_huge_pte(pte_t entry, struct vm_area_struct *vma,
 
 static inline struct hstate *page_hstate(struct page *page)
 {
+ VM_BUG_ON(!PageHuge(page));
  return size_to_hstate(PAGE_SIZE << compound_order(page));
 }
 
-- 
1.7.1